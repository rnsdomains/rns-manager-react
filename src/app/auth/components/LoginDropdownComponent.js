import React, { useState, useEffect } from 'react';
import propTypes from 'prop-types';
import { multilanguage } from 'redux-multilanguage';
import { Button } from 'react-bootstrap';

import LoginFormContainer from '../containers/LoginFormContainer';
import SingleDomainComponent from './SingleDomainComponent';

const LoginDropDownComponent = ({
  strings, name, handleLogin, authError, getPreviousDomains, isLoggedIn, isWalletConnected,
  showPopUp, toggleShowPopUp, disconnectDomain, disconnectWallet, redirectAdmin,
}) => {
  const [previousDomains, setPreviousDomains] = useState([]);
  useEffect(() => {
    if (showPopUp) {
      setPreviousDomains(getPreviousDomains());
    }
  }, [showPopUp]);

  let buttonText;
  if (!isWalletConnected) {
    buttonText = strings.connect_wallet;
  } else {
    buttonText = isLoggedIn ? name : strings.login;
  }

  const handleDisconnectClick = (domain) => {
    disconnectDomain(domain);
    setPreviousDomains(previousDomains.filter(d => (d.domain !== domain)));
  };
  const handleLoginClick = domain => handleLogin(domain);

  return (
    <div className="loginDropdown nav-item">
      <Button className="start" onClick={toggleShowPopUp}>
        {buttonText}
      </Button>

      {showPopUp
      && (
        <ul className="popup">
          {isLoggedIn && (
            <SingleDomainComponent
              domain={name}
              handleDisconnectClick={handleDisconnectClick}
              handleTextClick={redirectAdmin}
              isCurrent
            />
          )}

          {previousDomains.map(addr => (
            <SingleDomainComponent
              key={addr.domain}
              domain={addr.domain}
              handleDisconnectClick={handleDisconnectClick}
              handleTextClick={handleLoginClick}
            />
          ))}

          <LoginFormContainer
            authError={authError}
            handleLogin={handleLoginClick}
            showLoginInitState={(!isLoggedIn && previousDomains.length === 0) || authError}
          />

          <li className="disconnect">
            <button type="button" onClick={disconnectWallet}>{strings.disconnect_wallet}</button>
          </li>
        </ul>
      )}
    </div>
  );
};

LoginDropDownComponent.defaultProps = {
  name: null,
};

LoginDropDownComponent.propTypes = {
  strings: propTypes.shape({
    login: propTypes.string.isRequired,
    your_domain: propTypes.string.isRequired,
    enter: propTypes.string.isRequired,
    disconnect_wallet: propTypes.string.isRequired,
    not_domains_owner_message: propTypes.string.isRequired,
    connect_wallet: propTypes.string.isRequired,
  }).isRequired,
  name: propTypes.string,
  handleLogin: propTypes.func.isRequired,
  authError: propTypes.bool.isRequired,
  showPopUp: propTypes.bool.isRequired,
  toggleShowPopUp: propTypes.func.isRequired,
  getPreviousDomains: propTypes.func.isRequired,
  disconnectDomain: propTypes.func.isRequired,
  disconnectWallet: propTypes.func.isRequired,
  redirectAdmin: propTypes.func.isRequired,
  isLoggedIn: propTypes.bool.isRequired,
  isWalletConnected: propTypes.bool.isRequired,
};

export default multilanguage(LoginDropDownComponent);
