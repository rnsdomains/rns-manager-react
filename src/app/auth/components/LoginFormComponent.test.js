import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import LoginDropdownComponent from './LoginFormComponent';
import mockStore from '../../../../tests/config/mockStore';

const store = mockStore({
  not_domains_owner_message: 'You are not the owner',
  enter: 'enter',
  your_domain: 'your domain',
  add_account: 'add account',
  invalid_name: 'invalid name',
});

describe('LoginDropdownComponent', () => {
  const initProps = {
    authError: false,
    showLoginInitState: false,
    handleLogin: jest.fn(),
    domainInputInitialState: '',
  };

  const generateComponent = (localProps = {}) => {
    const combinedProps = { ...initProps, ...localProps };
    return mount(
      <Provider store={store}>
        <LoginDropdownComponent {...combinedProps} />
      </Provider>,
    );
  };

  it('renders and matches snapshot when closed', () => {
    const wrapper = generateComponent();
    expect(wrapper.find('button').text()).toBe('+ add account');
  });

  it('loads the initial state of the input box', () => {
    const wrapper = generateComponent({ showLoginInitState: true, domainInputInitialState: 'foobar' });
    expect(wrapper.find('input').props().value).toBe('foobar');
  });

  it('shows error when there is one', () => {
    const wrapper = generateComponent({ showLoginInitState: true, authError: true });
    expect(wrapper.find('.error').text()).toBe('You are not the owner');
  });

  describe('login events', () => {
    it('sends the domain when form is submitted', () => {
      const handleLogin = jest.fn();
      const wrapper = generateComponent({ showLoginInitState: true, handleLogin });
      wrapper.find('input').simulate('change', { target: { value: 'hello' } });
      wrapper.find('button.btn').simulate('click');

      expect(handleLogin).toBeCalledWith('hello.rsk');
    });

    it('shows an error when domain is invalid', () => {
      const handleLogin = jest.fn();
      const wrapper = generateComponent({ showLoginInitState: true, domainInputInitialState: 'foobar!!', handleLogin });

      wrapper.find('button.btn').simulate('click');

      expect(wrapper.find('.error').text()).toBe('invalid name');
      expect(handleLogin).toBeCalledTimes(0);
    });
  });
});
