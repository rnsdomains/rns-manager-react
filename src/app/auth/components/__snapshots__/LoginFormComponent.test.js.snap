// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginDropdownComponent matches snapshot when open 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(LoginFormComponent)
    authError={false}
    handleLogin={[MockFunction]}
    showLoginInitState={true}
  >
    <LoginFormComponent
      authError={false}
      currentLanguageCode="en"
      dispatch={[Function]}
      handleLogin={[MockFunction]}
      showLoginInitState={true}
      strings={
        Object {
          "add_account": "Add account",
          "enter": "Enter",
          "not_domains_owner_message": "log out",
          "your_domain": "Your domain",
        }
      }
    >
      <div
        className="loginForm"
      >
        <h3>
          Your domain
        </h3>
        <div
          className="rskinput"
        >
          <input
            onChange={[Function]}
            value=""
          />
          <span>
            .rsk
          </span>
        </div>
        <Button
          active={false}
          disabled={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
          <button
            className="btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Enter
          </button>
        </Button>
      </div>
    </LoginFormComponent>
  </Connect(LoginFormComponent)>
</Provider>
`;

exports[`LoginDropdownComponent renders and matches snapshot when closed 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(LoginFormComponent)
    authError={false}
    handleLogin={[MockFunction]}
    showLoginInitState={false}
  >
    <LoginFormComponent
      authError={false}
      currentLanguageCode="en"
      dispatch={[Function]}
      handleLogin={[MockFunction]}
      showLoginInitState={false}
      strings={
        Object {
          "add_account": "Add account",
          "enter": "Enter",
          "not_domains_owner_message": "log out",
          "your_domain": "Your domain",
        }
      }
    >
      <div
        className="loginForm"
      >
        <Button
          active={false}
          className="showLogin"
          disabled={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
          <button
            className="showLogin btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            + 
            Add account
          </button>
        </Button>
      </div>
    </LoginFormComponent>
  </Connect(LoginFormComponent)>
</Provider>
`;
